source/opt/decoration_manager.cpp

```diff
@@ -231,29 +231,78 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
 
 void DecorationManager::ForEachDecoration(uint32_t id,
                                           uint32_t decoration,
-                                          std::function<void(const ir::Instruction&)> f) const {
-  auto decoration_list = id_to_decoration_insts_.find(id);
-  if (decoration_list != id_to_decoration_insts_.end()) {
-    for (const ir::Instruction* inst : decoration_list->second) {
-      switch (inst->opcode()) {
-        case SpvOpDecorate:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, true)) {
```

Comment:
I see why you did this.  You wanted to expose the effective decorations.  The group decorations were getting in the way.

Please clarify the method comment for this method. (i.e. in the header) to indicate this intent.

It's not your problem, but it's unfortunate that GetDecorationsFor allocates a new vector that we then throw away.  I suspect a good future enhancement is to have a ForEachDecoration overload that doesn't constrain the decoration kind.  This method could be written in terms of that more general one.